src/Workspaces/Core/Portable/Utilities/EventMap.cs

```diff
@@ -18,64 +19,133 @@ public EventMap()
         }
 
         public void AddEventHandler<TEventHandler>(string eventName, TEventHandler eventHandler)
+            where TEventHandler : class
         {
             using (_guard.DisposableWait())
             {
-                var handlers = GetEvents_NoLock<TEventHandler>(eventName);
-                var newHandlers = handlers.Add(eventHandler);
-                SetEvents_NoLock(eventName, newHandlers);
+                var registries = GetRegistries_NoLock<TEventHandler>(eventName);
+                var newHandlers = registries.Add(new Registry<TEventHandler>(eventHandler));
+                SetRegistries_NoLock(eventName, newHandlers);
             }
         }
 
         public void RemoveEventHandler<TEventHandler>(string eventName, TEventHandler eventHandler)
+            where TEventHandler : class
         {
             using (_guard.DisposableWait())
             {
-                var handlers = GetEvents_NoLock<TEventHandler>(eventName);
-                var newHandlers = handlers.Remove(eventHandler);
-                if (newHandlers != handlers)
+                var registries = GetRegistries_NoLock<TEventHandler>(eventName);
+
+                // remove disabled registrations from list
+                var newRegistries = registries.RemoveAll(r => r.HasHandler(eventHandler));
+
+                if (newRegistries != registries)
                 {
-                    SetEvents_NoLock(eventName, newHandlers);
+                    // disable all registrations of this handler (so pending raise events can be squelched)
+                    // This does not guarantee no race condition between Raise and Remove but greatly reduces it.
```

Comment:
There is always a race for multi-threaded apps, but this closes the race completely for apps with an STA because both the Remove and the Raise will happen on that same thread through the synchronization context?